src/main/java/com/hazelcast/aws/impl/AwsOperation.java

```diff
@@ -0,0 +1,309 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.TimeZone;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.aws.utility.StringUtils.isEmpty;
+import static com.hazelcast.aws.utility.StringUtils.isNotEmpty;
+import static com.hazelcast.nio.IOUtil.closeResource;
+
+/**
+ * Abstract base class for AWS operations returning a response object of type E.
+ * Used by AwsClientStrategy implementations for calling AWS service endpoints.
+ */
+public abstract class AwsOperation<E> {
+
+    /**
+     * URI to fetch container credentials (when IAM role is enabled)
+     * <p>
+     * see http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html
+     */
+    static final String IAM_TASK_ROLE_ENDPOINT = "http://169.254.170.2";
+
+    static final String UTF8_ENCODING = "UTF-8";
+
+    private static final ILogger LOGGER = Logger.getLogger(AwsOperation.class);
+    private static final int MIN_HTTP_CODE_FOR_AWS_ERROR = 400;
+    private static final int MAX_HTTP_CODE_FOR_AWS_ERROR = 600;
+
+    protected final String docVersion;
+
+    final AwsConfig awsConfig;
+    final AwsCredentials awsCredentials;
+    final Map<String, String> attributes = new HashMap<String, String>();
+    final Map<String, String> headers = new HashMap<String, String>();
+
+    String body = "";
+
+    private final URL endpointURL;
+    private final String service;
+    private final String httpMethod;
+
+    /**
+     * Creates an AwsOperation
+     * @param awsConfig configuration
+     * @param endpointURL endpoint URL
+     * @param service AWS service name, e.g., "ec2"
+     * @param docVersion service document version (see AWS API docs for service)
+     * @param httpMethod HTTP method name to use for this operation, e.g., "POST"
+     */
+    AwsOperation(AwsConfig awsConfig, URL endpointURL, String service, String docVersion, String httpMethod) {
+        this.awsConfig = awsConfig;
+        this.awsCredentials = new AwsCredentials(awsConfig);
+        this.endpointURL = endpointURL;
+        this.service = service;
+        this.docVersion = docVersion;
+        this.httpMethod = httpMethod;
+    }
+
+    /**
+     * Invokes this service, unmarshal the response and return it.
+     *
+     * @param args service arguments
+     * @return the response
+     * @throws Exception if there is an exception invoking the service
+     */
+    public E execute(Object... args) throws Exception {
+        if (isNotEmpty(awsCredentials.getIamRole()) || isEmpty(awsCredentials.getAccessKey())) {
+            retrieveCredentials();
+        }
+
+        LOGGER.finest("OK we have credentials, signing request...");
```

Comment:
Could you make this logging more user-friendly?

Reply:
Noted.